ArduCopter/mode_zigzag.cpp

```diff
@@ -0,0 +1,277 @@
+    } else {
+        // clear out pilot desired acceleration in case radio failsafe event occurs and we
+        // do not switch to RTL for some reason
+        loiter_nav->clear_pilot_desired_acceleration();
+    }
+
+    // set motors to full range
+    motors->set_desired_spool_state(AP_Motors::DESIRED_THROTTLE_UNLIMITED);
+
+    // run loiter controller
+    loiter_nav->update(ekfGndSpdLimit, ekfNavVelGainScaler);
+
+    // call attitude controller
+    attitude_control->input_euler_angle_roll_pitch_euler_rate_yaw(loiter_nav->get_roll(), loiter_nav->get_pitch(), target_yaw_rate);
+
+    // adjust climb rate using rangefinder
+    target_climb_rate = get_surface_tracking_climb_rate(target_climb_rate, pos_control->get_alt_target(), G_Dt);
+
+    // get avoidance adjusted climb rate
+    target_climb_rate = get_avoidance_adjusted_climbrate(target_climb_rate);
+
+    // update altitude target and call position controller
+    pos_control->set_alt_target_from_climb_rate_ff(target_climb_rate, G_Dt, false);
+
+    // adjusts target up or down using a climb rate
+    pos_control->update_z_controller();
+}
+
+// return true if vehicle is within a small area around the destination
+bool Copter::ModeZigZag::reached_destination()
+{
+    // check if wp_nav believes it has reached the destination
+    if (!wp_nav->reached_wp_destination()) {
+        return false;
+    }
+
+    // check distance to destination
+    if (wp_nav->get_wp_distance_to_destination() > ZIGZAG_WP_RADIUS_CM) {
+        return false;
+    }
+
+    // wait at least one second
+    uint32_t now = AP_HAL::millis();
+    if (reach_wp_time_ms == 0) {
+        reach_wp_time_ms = now;
+    }
+    return ((now - reach_wp_time_ms) > 1000);
+}
+
+// calculate next destination according to vector A-B and current position
+bool Copter::ModeZigZag::calculate_next_dest(uint8_t dest_num, Vector3f& next_dest) const
+{
+    // sanity check dest_num
+    if (dest_num > 1) {
+        return false;
+    }
+
+    // define start_pos as either A or B depending upon dest_num
+    Vector2f start_pos = dest_num == 0 ? dest_A : dest_B;
+
+    // calculate vector from A to B
+    Vector2f AB_diff = dest_B - dest_A;
+
+    // check distance between A and B
+    if (!is_positive(AB_diff.length_squared())) {
+        return false;
+    }
+
+    // get distance from vehicle to start_pos
+    const Vector3f curr_pos = inertial_nav.get_position();
+    const Vector2f curr_pos2d = Vector2f(curr_pos.x, curr_pos.y);
+    Vector2f veh_to_start_pos = curr_pos2d - start_pos;
+
+    // lengthen AB_diff so that it is at least as long as vehicle is from start point
+    // we need to ensure that the lines perpendicular to AB are long enough to reach the vehicle
+    float scalar = 1.0f;
+    if (veh_to_start_pos.length_squared() > AB_diff.length_squared()) {
+        scalar = veh_to_start_pos.length() / AB_diff.length();
```

Comment:
we should add a protection on dest_B != dest_A to prevent divide by zero. It would be easy to trigger with a switch

Reply:
@khancyr, txs for the check.  There's actually a check higher up in the function that AB_diff.length_squared() is positive but still, I will add a check up when we store the A and B points so it doesn't advance out of the STORING_POINTS stage unless A and B are different.